EIPS/eip-7432.md

```diff
@@ -0,0 +1,346 @@
+* The `hasRole` function MAY be implemented as `pure` or `view`.
+* The `hasRole` function SHOULD return `false` if `_supportsMultipleAssignments` is `false` and last role assignment is
+not to `_grantee` (see [Unique and Non-Unique Roles](#unique-and-non-unique-roles) for more).
+* The `roleData` function MAY be implemented as `pure` or `view`.
+* Compliant contracts SHOULD support [ERC-165](./eip-165.md).
+
+## Rationale
+
+[ERC-7432](./eip-7432.md) IS NOT an extension of [ERC-721](./eip-721.md) or [ERC-1155](./eip-1155.md). The main reason
+behind this decision is to keep the standard agnostic of any NFT implementation. This approach also enables the standard
+to be implemented externally or on the same contract as the NFT, and allow dApps to use roles with immutable NFTs.
+
+### Automatic Expiration
+
+Automatic expiration is implemented via the `grantRole` and `hasRole` functions. `grantRole` is responsible for setting
+the expiration date, and `hasRole` checks if the role is expired by comparing with the current block timestamp
+(`block.timestamp`). Since `uint256` is not natively supported by most programming languages, dates are represented as
+`uint64` on this standard. The maximum UNIX timestamp represented by a `uint64` is about the year `584,942,417,355`,
+which should be enough to be considered "permanent". For this reason, it's RECOMMENDED using `type(uint64).max` when
+calling the `grantRole` function to support use cases that require an assignment never to expire.
+
+### Unique and Non-Unique Roles
+
+The standard supports both unique and non-unique roles. Unique roles are roles that can be assigned to only one account,
+while non-unique roles can be granted to multiple accounts simultaneously. The parameter `_supportsMultipleAssignments`
+was included in the `hasRole` function to support both use cases. When `_supportsMultipleAssignments` is `true`, the
+function checks if the assignment exists and is not expired. However, when `false`, the function also validates that no
+other role was granted afterward. In other words, for unique roles, each new assignment invalidates the previous one,
+and only the last one can be valid.
+
+Assuming that the **role was granted and is not expired**, the following table shows the result the `hasRole` function
+MUST return:
+
+|     Role Type     | `_supportsMultipleAssignments` |  Is last Role granted?  | Result of `hasRole` |
+|:-----------------:|:------------------------------:|:-----------------------:|:-------------------:|
+|  Non-Unique Role  |            `false`             |       Irrelevant        |       `true`        |
+|    Unique Role    |             `true`             |         `true`          |       `true`        |
+|    Unique Role    |             `true`             |         `false`         |       `false`       |
+
+In conclusion, the `_supportsMultipleAssignments` argument only affects the result when `true`, and if the queried
+assignment is not the last one granted.
```

Comment:
This belongs in the specification section, since it defines behaviour. In rationale, for example, you might want to explain _why_ you support unique and non-unique roles, and why the uniqueness boolean is set when checking the role instead of when assigning it.

Reply:
As you suggested over the call, we split the `hasRole` function in two: `hasRole` and `hasUniqueRole`. This also removed the `_supportsMultipleAssignments` which greatly simplified the rationale/specification